Encrypted Signing Requests/Responses

vzotova opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
- Depends on https://github.com/nucypher/nucypher-core/pull/116
- Depends on https://github.com/nucypher/nucypher-contracts/pull/438
- Based over https://github.com/nucypher/nucypher/pull/3646

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

vzotova committed:
Prepare ground for signing request encryption

derekpierre committed:
Utilize encrypted signing request/responses from `nucypher-core`.

vzotova committed:
better SigningRequestDecryptingPower code

vzotova committed:
Adds newsfragment

derekpierre commented on nucypher/crypto/powers.py:
```diff
@@ -460,6 +465,8 @@ def _get_static_secret_from_ritual_id(self, ritual_id: int) -> SessionStaticSecr
     def get_pubkey_from_ritual_id(self, ritual_id: int) -> SessionStaticKey:
```
Let's rename `_get_static_secret_from_ritual_id` and `get_pubkey_from_ritual_id` to something more generic since signing uses `cohort_id` instead of `ritual_id`

Something like `_get_static_secret_from_id`/`get_pubkey_from_id` or ...?

vzotova replied:
✔️

derekpierre commented on nucypher/crypto/powers.py:
```diff
@@ -490,6 +497,38 @@ def encrypt_decryption_response(
+    def decrypt_encrypted_request(
+        self,
+        encrypted_request: EncryptedThresholdDecryptionRequest,  # EncryptedThresholdSigningRequest
+    ) -> ThresholdDecryptionRequest:  # ThresholdSigningRequest:
```
Ah, I see the short coming now of waiting for my `nucypher-core` changes... I missed this need. Perhaps don't put any typehints for now? (Same for the encrypt_decryption_response method below)

wdyt?

derekpierre commented on nucypher/crypto/powers.py:
```diff
@@ -490,6 +497,38 @@ def encrypt_decryption_response(
             raise self.ThresholdResponseEncryptionFailed from e
 
 
+class SigningRequestDecryptingPower(ThresholdRequestPower):
```
Let's leave "Decrypting" out of it so signing is separate from decryption. It's a little confusing with decryption in the name. wdyt?
```suggestion
class ThresholdSigningRequestPower(ThresholdRequestPower):
```

vzotova replied:
✔️

derekpierre commented on nucypher/crypto/powers.py:
```diff
@@ -460,6 +465,8 @@ def _get_static_secret_from_ritual_id(self, ritual_id: int) -> SessionStaticSecr
         return self._get_static_secret_from_ritual_id(ritual_id).public_key()
 
+
+class ThresholdRequestDecryptingPower(ThresholdRequestPower):
```
Let's rename to match format of Signing power?

```suggestion
class ThresholdDecryptionRequestPower(ThresholdRequestPower):
```

vzotova replied:
✔️

derekpierre reviewed (commented):
I totally missed that you need Objects (encrypted request/responses) to take as parameters or return from the methods for `ThresholdSigningRequestPower`. 

Makes me wonder if we should instead combine step (1) and step (2) (https://github.com/nucypher/nucypher-contracts/pull/438#discussion_r2481306997) into one comprehensive PR instead. Feels that way.

vzotova committed:
Renames of request powers

derekpierre committed:
Combine power creation for Signing/DecryptingRequestPower into one code block since they both share a common parent class.

derekpierre committed:
Add documentation for the various powers.

derekpierre committed:
Add Optional to typehint for signer parameter for ThresholdSigningPower since it can be None.

derekpierre committed:
Update SigningRequestPower to use `nucypher-core` objects for handling signing requests/responses.

derekpierre committed:
Update handling of signing request to leverage signing request power for decrypting/encrypting signing requests/responses.

derekpierre committed:
Move handling of threshold decryption requests from Ursula to Operator.
Now that it is moved previous functions exposed by Operator are no longer needed.

derekpierre committed:
Appease linter.

derekpierre committed:
Add keystore tests to ensure that recreated instance of keystore produces the same values for the new signing/decrypting request powers.

derekpierre committed:
Add unit tests for Signing/Decryping request powers.

derekpierre committed:
Use latest `nucypher-core` commit.

derekpierre committed:
Use updated feature newsfragment for #3666.

derekpierre commented on pyproject.toml:
```diff
@@ -9,7 +9,7 @@ python = ">=3.10,<4"
 ckzg = ">=2.1.1"  # required for python 3.13 support
 
-nucypher-core = "==0.15.0"
+nucypher-core = { git = "https://github.com/derekpierre/nucypher-core.git", branch = "signing", subdirectory = "nucypher-core-python" }
```
Leaving for now until code is merged.

derekpierre commented on tests/acceptance/ape-config.yaml:
```diff
@@ -5,8 +5,8 @@ plugins:
-    github: nucypher/nucypher-contracts
-    ref: signing
+    github: vzotova/nucypher-contracts
+    ref: request-key
```
Leaving for now until relevant PR is merged.

derekpierre committed:
Fix misspelling.

derekpierre commented on nucypher/characters/lawful.py:
```diff
@@ -698,13 +701,24 @@ def request_threshold_signatures(
+        requester_public_key = requester_sk.public_key()
+
+        shared_secrets = {}
+        encrypted_signing_rquests = {}
```
✔️

derekpierre committed:
Install git during docker build so that any git dependencies can be installed; this is helpful for development builds eg. needing to use unreleased `nucypher-core` code.

derekpierre committed:
Remove lingering debugging traceback code.

derekpierre committed:
Return BAD_REQUEST when there is an issue processing the encrypted request.

derekpierre reviewed (approved):
My approval is implicit since I'm a co-author of the PR.

manumonti reviewed (approved):
LGTM!

derekpierre committed:
Update `nucypher-core` dependency to latest commit on `main`.
Update ape-config.yaml to use latest `nucypher-contracts` on `main` now that associated PR is merged.
